pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/NamespaceOperation.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pulsar.common.policies.data;
+
+/**
+ * Namespace authorization operations.
+ */
+public enum NamespaceOperation {
```

Comment:
In my opinion we should reduce these operations otherwise things will get out of hand. Any SET_ could just be Update or Write. Any GET_ could just be Read or Get. Same for the others below. My concern is things will get out of hand and these lists will be very difficult to maintain. Is there a way to group the operations to reduce the number? These are just needed for authorization right?

Reply:
I added these operations from PIP-49 (https://github.com/apache/pulsar/wiki/PIP-49%3A-Permission-levels-and-inheritance), but I'm currently reworking all. As you can see (for example), `TenantOperation` has `DELETE_NAMESPACES` which is not in `NamespaceOperation`. Any resource operation should be managed/authorized by the resource which owns it.

`ClusterOperations` should manage `Tenants`, `TenantOperation` should manage `Namespaces`, etc. Hierarchical approach.

Reply:
I think we can simply these actions. I don't understand why an action is needed for every rest endpoint. For example, you can either update a namespace or not. You can either read a namespace or not. Why do you need fine grained control for updating/reading each field in a namespace? A simple Write permission can be applied to any action that changes a namespace. 

In my opinion, having an action for every endpoint makes the code very difficult to change and it's hard to maintain. How will you depreciate actions and provide backward compatibility? What happens when you add a new field to a namespace policy? This could break exiting implementations and their permissions or force users to reconfigure authorization rules. 

I see 2 basic actions needed to start for namespaces:
- Write
- Read

Same for the most of the other entities. We should start with a small set. These actions will become public api and will be very difficult to change or modify.

I disagree with the layout and operation mapping from PIP-49. 

I agree with @sijie and think we should simplify the verbs and I think pulsar should move more towards a policy/role based system like nomad, consul, vault, and or kubernetes.

Reply:
We can probably simplify these actions I totally agree, I followed the validated PIP-49. Simplify them is really hard: for example, list Namespaces in a tenant should be tenant level authz but only for list Namespaces because it can be done by a tenant user. But set rates for tenant must only be done by super User and not a user, so we need these split.

I guess we can have this issue a lot of times, not every endpoints, but it's hard to maintain. Which endpoint goes for this operation or this operation. Split all of them if easy to maintain. This comes with the defaultauthz provider which keeps the default role based as super user, tenant admin, or producer/consume authz. But the provider is passing the operations.

The goal here is to enable plugins authz to have all authz powers and all the context to authorize. In default case role is a user, but when you consider role as a token with all authorized info deserialized in custom authz provider it's so powerful.

Reply:
Just rereading my words, and the example I gave is read/write as you mentioned, my bad. I just don't want to limit authz possibilities. And I'm not sure that Read/write cover all cases.

Reply:
I think this mixes the operations on a namespace and the operations on a namespace policy. I would suggest separating them.